torch_np/_wrapper.py

```diff
@@ -179,14 +177,28 @@ def logspace(start, stop, num=50, endpoint=True, base=10.0, dtype=None, axis=0):
     return asarray(torch.logspace(start, stop, num, base=base, dtype=dtype))
 
 
-def arange(start, stop=None, step=1, dtype=None, *, like=None):
+def arange(start=None, stop=None, step=1, dtype=None, *, like=None):
     _util.subok_not_ok(like)
+    if step == 0:
+        raise ZeroDivisionError
+    if stop is None and start is None:
+        raise TypeError
     if stop is None:
-        # arange(stop)
-        stop = start
+        # XXX: this breaks if start is passed as a kwarg:
+        # arange(start=4) should raise (no stop) but doesn't
+        start, stop = 0, start
```

Comment:
It's a funny little glitch how both pytorch and numpy `arange` signatures cannot be easily represented in pure python without going fully manual on `arange(*args, **kwargs)`. For instance, https://pytorch.org/docs/stable/generated/torch.arange.html lists `torch.arange(start=0, end, step=1, *, ...)` which is not valid python (`end` does not have a default). While this breaks at least one numpy test, I'd rather leave it as is :-).

Reply:
Your approach goes in the right direction, but we may be able to do better?
https://github.com/pytorch/pytorch/blob/2b52db9c953d063db7b46c12f4df35b47aca4381/torch/_refs/__init__.py#L4221-L4237

Reply:
Applying this patch

```
(pytorch) br@gonzales:~/sweethome/proj/scipy/torch_np_compat$ git diff
diff --git a/torch_np/_wrapper.py b/torch_np/_wrapper.py
index bea4c2f..82444d5 100644
--- a/torch_np/_wrapper.py
+++ b/torch_np/_wrapper.py
@@ -177,18 +177,20 @@ def logspace(start, stop, num=50, endpoint=True, base=10.0, dtype=None, axis=0):
     return asarray(torch.logspace(start, stop, num, base=base, dtype=dtype))
 
 
-def arange(start=None, stop=None, step=1, dtype=None, *, like=None):
+def arange(start=0, stop=None, step=1, dtype=None, *, like=None):
     _util.subok_not_ok(like)
     if step == 0:
         raise ZeroDivisionError
-    if stop is None and start is None:
-        raise TypeError
-    if stop is None:
-        # XXX: this breaks if start is passed as a kwarg:
+#    if stop is None and start is None:
+#        raise TypeError
+#    if stop is None:
+#        # XXX: this breaks if start is passed as a kwarg:
         # arange(start=4) should raise (no stop) but doesn't
+#        start, stop = 0, start
+#    if start is None:
+#        start = 0
+    if stop is None:
         start, stop = 0, start
-    if start is None:
-        start = 0
``` 

results in

```
$ pytest torch_np/tests/test_function_base.py -v -x
====================================== test session starts =======================================
platform linux -- Python 3.9.0, pytest-7.2.0, pluggy-1.0.0 -- /home/br/miniconda3/envs/pytorch/bin/python3.9
cachedir: .pytest_cache
rootdir: /home/br/sweethome/proj/scipy/torch_np_compat
plugins: cov-4.0.0
collected 12 items                                                                               

torch_np/tests/test_function_base.py::TestCopy::test_basic PASSED                          [  8%]
torch_np/tests/test_function_base.py::TestCopy::test_order SKIPPED (ndarray.flags not ...) [ 16%]
torch_np/tests/test_function_base.py::TestArange::test_infinite PASSED                     [ 25%]
torch_np/tests/test_function_base.py::TestArange::test_nan_step PASSED                     [ 33%]
torch_np/tests/test_function_base.py::TestArange::test_zero_step PASSED                    [ 41%]
torch_np/tests/test_function_base.py::TestArange::test_require_range FAILED                [ 50%]

============================================ FAILURES ============================================
_________________________________ TestArange.test_require_range __________________________________

self = <torch_np.tests.test_function_base.TestArange object at 0x7f4655d33580>

    def test_require_range(self):
>       assert_raises(TypeError, np.arange)
E       Failed: DID NOT RAISE <class 'TypeError'>

torch_np/tests/test_function_base.py:60: Failed
```

where the test is straight copy-paste from numpy.

Reply:
Another way of doing this is via overloads. you can see in `native_functions.yaml` the overloads that we have in C++ for this function. It was done like this before in primtorch, but I "fixed it" (clearly not perfectly, as you've found) at some point. The previous implementation went something like
https://github.com/pytorch/pytorch/blob/2f95b61ceaab65580d8dece821693d8b31ecfca8/torch/_refs/__init__.py#L3113-L3138

Reply:
It's pretty clear that these are C++ overloads, but this and `dim=..., keepdims=...`. And that they do not cleanly map onto the cartesian product behavior of python's arguments with defaults, unless without a borderline unreasonable amount of effort anyway.

With arange specifically, I'm with you FWIW. Doing `if stop is None: start, stop = 0, start` is close to the only reasonable thing to do. 
When I saw `arange(start=5)` in numpy tests, my first reaction was along the lines of _what in the name of all things graceful is this_.

Anyway, I don't believe this is worth getting to the bottom of, as long as we have better things to do. 

The main take-home message for myself is that positional-only args in python are a thing, since `arange(start, /, stop=None, step=1)` would sidestep this all completely.

Reply:
also, note that, as discussed, implementing a superset of NumPy's functionality is also alright, so in this case we should be good :)